### PR TITLE
entity: added model + api

### PR DESCRIPTION
**Pull request by manuelpolzhofer**
- added entity model
- added tests for entity model
- defined rest api Protobuf for model (needed to implement model init methods)

### CODE REVIEW

File: documents/entity/model_test.go [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+	assert.True(t, errors.IsOfType(documents.ErrDocumentInvalidType, err))
+	assert.NoError(t, testRepo().Create(id1[:], entity.CurrentVersion(), entity))
+
+	// update the document
+	model, err := testRepo().Get(id1[:], entity.CurrentVersion())
+	assert.NoError(t, err)
+	oldEntity := model.(*Entity)
+	data := oldEntity.getClientData()
+	data.LegalName = "new legal name"
+	err = entity.PrepareNewVersion(entity, data, []string{id3.String()})
```
Comment (manuelpolzhofer) [inline documents/entity/model_test.go]: @vedhavyas I am adding `id3` as a collaborator and by default, he has write-access to everything?

File: Gopkg.toml [hunk 2]
```diff
@@ -28,7 +28,8 @@ required = ["github.com/centrifuge/centrifuge-ethereum-contracts", "github.com/r
 
 [[constraint]]
   name = "github.com/centrifuge/centrifuge-protobufs"
-  revision = "ef4b1ce4b74db439613563007c9d0dfd00c54e41"
+  source = "github.com/manuelpolzhofer/centrifuge-protobufs"
```
Comment (mikiquantum) [inline Gopkg.toml]: 👍

File: Makefile [hunk 3]
```diff
@@ -55,7 +55,7 @@ proto-gen-go: ## generates the go bindings
 proto-all: ## runs prototool all
 	$(PROTOTOOL_BIN) all
 
-gen-swagger: ## generates the swagger documentation
+    gen-swagger: ## generates the swagger documentation
```
Comment (mikiquantum) [inline Makefile]: Is this just GH formatting ?

File: documents/entity/model.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+// CreateNFTProofs creates proofs specific to NFT minting.
+func (e *Entity) CreateNFTProofs(
+	account identity.DID,
+	registry common.Address,
+	tokenID []byte,
+	nftUniqueProof, readAccessProof bool) (proofs []*proofspb.Proof, err error) {
+	return e.CoreDocument.CreateNFTProofs(
+		e.DocumentType(),
```
Comment (charleenfei) [inline documents/entity/model.go]: tree?

File: documents/entity/model.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
+package entity
+
+import (
+	"encoding/json"
+	"reflect"
+
+	cliententitypb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/entity"
+
```
Comment (charleenfei) [inline documents/entity/model.go]: import spacing

File: documents/entity/model.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+const prefix string = "entity"
+
+// tree prefixes for specific to documents use the second byte of a 4 byte slice by convention
+func compactPrefix() []byte { return []byte{0, 3, 0, 0} }
+
+// Entity implements the documents.Model keeps track of entity related fields and state
+type Entity struct {
+	*documents.CoreDocument
+
+	Identity  []byte
```
Comment (mikiquantum) [inline documents/entity/model.go]: can we use here:
```suggestion
	Identity  *identity.DID
```

File: documents/entity/model_test.go [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/centrifuge/centrifuge-protobufs/gen/go/entity"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/identity"
+	cliententitypb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/entity"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
```
Comment (charleenfei) [inline documents/entity/model_test.go]: import spacing

File: documents/entity/model.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+		return errors.New("failed to init core document: %v", err)
+	}
+
+	e.CoreDocument = cd
+	return nil
+}
+
+// initEntityFromData initialises entity from entityData
+func (e *Entity) initEntityFromData(data *cliententitypb.EntityData) error {
+	e.Identity = data.Identity
```
Comment (mikiquantum) [inline documents/entity/model.go]: We shouldnt expose byte[] in client proto, so there will be extra processing here, to convert to model.

File: protobufs/entity/service.proto [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  string document_id = 1;
+  string version_id = 2;
+  string state = 3;
+  repeated string collaborators = 4;
+  string transaction_id = 5;
+}
+
+// EntityData is the default entity schema
+message EntityData {
+  bytes identity = 1;
```
Comment (mikiquantum) [inline protobufs/entity/service.proto]: Should be string

File: protobufs/gen/swagger/fuuu/entity.swagger.json [hunk 10]
```diff
@@ -0,0 +1,366 @@
+{
```
Comment (mikiquantum) [inline protobufs/gen/swagger/fuuu/entity.swagger.json]: What is this?? `fuuu`

File: documents/entity/bootstrapper.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
+package entity
+
+/*
```
Comment (mikiquantum) [inline documents/entity/bootstrapper.go]: Why commented out?

File: documents/entity/bootstrapper_test.go [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+	"testing"
+)
+
+func TestBootstrapper_Bootstrap(t *testing.T) {
+	//err := (&Bootstrapper{}).Bootstrap(map[string]interface{}{})
+	//assert.Error(t, err, "Should throw an error because of empty context")
+}
+
+func TestBootstrapper_registerEntityService(t *testing.T) {
+	//context := map[string]interface{}{}
```
Comment (mikiquantum) [inline documents/entity/bootstrapper_test.go]: Why commented out?

File: documents/entity/model.go (hunk 6 above)
Comment (manuelpolzhofer) [reply documents/entity/model.go]: 👍

File: protobufs/gen/swagger/fuuu/entity.swagger.json (hunk 10 above)
Comment (manuelpolzhofer) [reply protobufs/gen/swagger/fuuu/entity.swagger.json]: :) leftover from trying out stuff. I will remove it. good catch

**Comment by manuelpolzhofer** [comment]
@mikiquantum  addressed your issues

File: documents/entity/bootstrapper.go (hunk 11 above)
Comment (manuelpolzhofer) [reply documents/entity/bootstrapper.go]: Waiting for the service to activate it

File: documents/entity/model.go (hunk 4 above)
Comment (mikiquantum) [reply documents/entity/model.go]: As @charleenfei pointed out, we need to pass the data tree here, right?

File: Makefile (hunk 3 above)
Comment (mikiquantum) [reply Makefile]: @manuelpolzhofer can u check why this formatting change?

File: documents/entity/model.go [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+	Contacts []*entitypb.Contact
+
+	EntitySalts *proofs.Salts
+}
+
+// getClientData returns the client data from the entity model
+func (e *Entity) getClientData() *cliententitypb.EntityData {
+	var didString string
+	if e.Identity != nil {
+		didString = hexutil.Encode(e.Identity[:])
```
Comment (vedhavyas) [inline documents/entity/model.go]: call `String` instead

File: documents/entity/model.go [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+		return errors.New("failed to init core document: %v", err)
+	}
+
+	e.CoreDocument = cd
+	return nil
+}
+
+// initEntityFromData initialises entity from entityData
+func (e *Entity) initEntityFromData(data *cliententitypb.EntityData) error {
+
```
Comment (vedhavyas) [inline documents/entity/model.go]: spacing

File: documents/entity/model.go [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+	if data.Identity != "" {
+		if did, err := identity.NewDIDFromString(data.Identity); err == nil {
+			e.Identity = &did
+		}
+	}
+	e.LegalName = data.LegalName
+	//e.Addresses = data.Addresses //TODO precise proofs not supporting fields yet
```
Comment (vedhavyas) [inline documents/entity/model.go]: can you add the unsupported fields to precise proofs and let @vimukthi-git know? Once his integration is done, it should be straightforward

File: documents/entity/model.go [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+	e.PaymentDetails = data.PaymentDetails
+	e.Contacts = data.Contacts
+
```
Comment (vedhavyas) [inline documents/entity/model.go]: spacing

File: documents/entity/model_test.go [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+	"github.com/stretchr/testify/mock"
+)
+
+var ctx = map[string]interface{}{}
+var cfg config.Configuration
+var configService config.Service
+var defaultDID = testingidentity.GenerateRandomDID()
+
+var (
+	cid         = testingidentity.GenerateRandomDID()
```
Comment (vedhavyas) [inline documents/entity/model_test.go]: `did`

File: documents/entity/model_test.go [hunk 18]
```diff
@@ -0,0 +1,382 @@
...
+	centIDBytes = cid[:]
```
Comment (vedhavyas) [inline documents/entity/model_test.go]: `didBytes`

File: documents/entity/model_test.go (hunk 1 above)
Comment (vedhavyas) [reply documents/entity/model_test.go]: yeap

File: documents/entity/model_test.go (hunk 17 above)
Comment (manuelpolzhofer) [reply documents/entity/model_test.go]: good catch

**Pull request merged**